include/lbann/execution_algorithms/training_algorithm.hpp

```diff
@@ -28,46 +28,122 @@
+ *  and the output is a trained model (or, a set of parameters that
+ *  define the action of the model). Here, "trained" means that the
+ *  training algorithm has evolved the parameters until user-specified
+ *  stopping criteria have been met; it does necessarily imply that
+ *  any underlying optimization method has converged (or even exists)
+ *  or that such a convergence is even well-defined.
+ *
+ *  A key capability is that training algorithms should be
+ *  composable. This allows metaheuristic algorithms to simply be
+ *  implemented as training algorithms constructed from one or more
+ *  "inner" training algorithms.
+ *
+ *  @todo One component that we need to address yet is the issue of
+ *        logically encapsulating multiple models, as either inputs or
+ *        outputs to a training algorithm. Specifically, consider the
+ *        LTFB "meta-learning" method. Rather than producing the
+ *        single best model, a user might be interested in the K best
+ *        models. In this case, tournament-based evolution will begin
+ *        with a single model (per trainer) but could output several
+ *        models. Similarly, one might begin with an arbitrary
+ *        collection of models that are evolved until a single best
+ *        model emerges. This draws in other issues to be addressed
+ *        elsewhere in LBANN such as "How do we export models?"
+ *        Currently, this is done by writing to files on disk via
+ *        callbacks. However, one might imagine "in-core" interation
+ *        between training and inference, perhaps in an online
+ *        learning scenario, in which repeatedly writing to and
+ *        reading from disk is not sufficient.
+ */
+class training_algorithm
+  : public Cloneable<HasAbstractFunction<training_algorithm>>
+{
 public:
+  /** @name Lifecycle Management */
+  ///@{
+  /** @brief Constructor
+   *  @param[in] name The user-defined name of the algorithm.
+   */
+  training_algorithm(std::string name);
+  virtual ~training_algorithm() = 0;
+  ///@}
 
-  /** Constructor. */
-  training_algorithm() {};
-  /** Copy constructor. */
-  training_algorithm(const training_algorithm& other) = default;
-  /** Copy assignment operator. */
-  training_algorithm& operator=(const training_algorithm& other) = default;
-  /** Move constructor. */
-  training_algorithm(training_algorithm&& other) = default;
-  /** Move assignment operator. */
-  training_algorithm& operator=(training_algorithm&& other) = default;
-  /** Destructor. */
-  virtual ~training_algorithm() = default;
-  /** Copy training_algorithm. */
-  //  virtual training_algorithm* copy() const = default;
-
-  virtual std::string get_name() const = 0;
+  /** @name Queries */
+  ///@{
+  /** @brief A string identifying the type of the object. */
+  virtual std::string get_type() const = 0;
+  /** @brief A user-defined string identifying the algorithm object. */
+  std::string const& get_name() const noexcept;
```

Comment:
Where is `get_name()` defined?

Reply:
Good catch. It got lost in a mess of git acrobatics.